projects/eval-pqmon/src/common/afe_config.c

```diff
@@ -0,0 +1,514 @@
+			    ADI_PQLIB_TOTAL_WAVEFORM_CHANNELS +
+			    2];
+
+	cmdBuffer[0] = (addr >> 4);
+	cmdBuffer[1] = ((addr & 0x0F) << 4) + 8;
+	memset((uint8_t *)&cmdBuffer[2], 0, sizeof(uint8_t) * numSamples * 4);
+	status =
+		no_os_spi_write_and_read(hSPI, (uint8_t *)cmdBuffer, 2 + numSamples * 4);
+
+	memcpy((uint8_t *)pData, &cmdBuffer[2], numSamples * 4);
+	no_os_memswap64((void *)pData, numSamples * 4, 4);
+
+	return status;
+}
+
+uint32_t afe_read_16bit_buff(uint16_t addr, uint16_t numSamples,
+			     uint16_t *pData)
+{
+	int32_t status = SYS_STATUS_SUCCESS;
+	uint8_t cmdBuffer[2 * ADI_PQLIB_WAVEFORM_BLOCK_SIZE *
+			    ADI_PQLIB_TOTAL_WAVEFORM_CHANNELS +
+			    2];
+
+	cmdBuffer[0] = (addr >> 4);
+	cmdBuffer[1] = ((addr & 0x0F) << 4) + 8;
+	memset((uint8_t *)&cmdBuffer[2], 0, sizeof(uint8_t) * numSamples * 2);
+	status =
+		no_os_spi_write_and_read(hSPI, (uint8_t *)cmdBuffer, 2 + numSamples * 2);
+
+	memcpy((uint8_t *)pData, &cmdBuffer[2], numSamples * 2);
+	no_os_memswap64((void *)pData, numSamples * 2, 2);
+
+	return status;
+}
+
+int32_t reset_afe(void)
+{
+	int32_t status = 0;
+	struct no_os_gpio_desc *sw_reset_gpio_desc;
+
+	status = no_os_gpio_get(&sw_reset_gpio_desc, &reset_gpio_ip);
+	if (status)
+		goto exit;
+	status = no_os_gpio_direction_output(sw_reset_gpio_desc, 1);
+	if (status)
+		goto exit;
+	status = no_os_gpio_set_value(sw_reset_gpio_desc, 0);
+	if (status)
+		goto exit;
+	no_os_mdelay(100);
+	status = no_os_gpio_set_value(sw_reset_gpio_desc, 1);
+	if (status)
+		goto exit;
+	no_os_mdelay(100);
+	return status;
+
+exit:
+	no_os_gpio_remove(sw_reset_gpio_desc);
+	return status;
+}
+
+int32_t spi_device_init(void)
+{
+	return no_os_spi_init(&hSPI, &spi_egy_ip);
+}
+
+int32_t afe_read_status0(uint32_t *pSTATUS0)
+{
+	int32_t status = SYS_STATUS_SUCCESS;
+
+	uint32_t status0 = 0;
+	*pSTATUS0 = 0;
+	if (rmsOneReady != 0) {
+		rmsOneReady = 0;
+		status = afe_read_32bit_buff(REG_STATUS0, 1, pSTATUS0);
+		status0 = *pSTATUS0;
+		if (status != 0) {
+			status = SYS_STATUS_AFE_STATUS0_FAILED;
+		}
+
+		if (status == 0) {
+			status0 |= BITM_STATUS0_RMSONERDY;
+			status = afe_write_32bit_reg(REG_STATUS0, (uint32_t *)&status0);
+			if (status != 0) {
+				status = SYS_STATUS_AFE_STATUS0_FAILED;
+			}
+		}
+	}
+
+	return status;
+}
+
+void afe_wait_settling(uint32_t cycles)
+{
+	uint32_t status0;
+	int32_t status = 0;
+	while (cycles > 0) {
+		if (rmsOneReady != 0) {
+			status = afe_read_status0((uint32_t *)&status0);
+			if ((status == SYS_STATUS_AFE_STATUS0_FAILED) ||
+			    (status == SYS_STATUS_SUCCESS)) {
+				cycles--;
+			}
+		}
+	}
+}
+
+int32_t afe_set_acc_mode(uint16_t frequency, uint16_t vconsel,
+			 uint16_t iconsel)
+{
+	int32_t status = SYS_STATUS_SUCCESS;
+	uint16_t config = 0;
+
+	vconsel <<= BITP_ACCMODE_VCONSEL;
+	iconsel <<= BITP_ACCMODE_ICONSEL;
+
+	if (frequency == 50) {
+		frequency = BITM_ACCMODE_SELFREQ_50HZ;
+	} else if (frequency == 60) {
+		frequency = BITM_ACCMODE_SELFREQ_60HZ;
+	}
```

Comment:
you probably need an else return -EINVAL here in case frequency is not 50 and not 60